File: csharp/go-counting/GoCounting.cs
public enum Owner
{
    None,
    Black,
    White
}

public class GoCounting
{
   private readonly char[,] board;
    private readonly int height;
    private readonly int width;
    
    public GoCounting(string input)
    {
        var lines = input.Split('\n', StringSplitOptions.RemoveEmptyEntries);
        height = lines.Length;
        width = lines[0].Length;
        board = new char[width, height]; //Note: board[x, y] (column, row)
        for (int y = 0; y < height; y++)
        {
            for (int x = 0; x < width; x++)
            {
                board[x, y] = lines[y][x];
            }
        }
    }

    public Tuple<Owner, HashSet<(int, int)>> Territory((int, int) coord)
    {
        var (x, y) = coord;
        if (!InBounds(x, y))
            throw new ArgumentException("Out of bounds");
        if (board[x, y] != ' ')
            return Tuple.Create(Owner.None, new HashSet<(int, int)>());
        var visited = new HashSet<(int, int)>();
        var queue = new Queue<(int, int)>();
        var borders = new HashSet<char>();
        queue.Enqueue((x, y));
        visited.Add((x, y));
        while (queue.Count > 0)
        {
            var (cx, cy) = queue.Dequeue();
            foreach (var (nx, ny) in Neighbors(cx, cy))
            {
                if (!InBounds(nx, ny) || visited.Contains((nx, ny)))
                    continue;
                var cell = board[nx, ny];
                if (cell == ' ')
                {
                    visited.Add((nx, ny));
                    queue.Enqueue((nx, ny));
                }
                else if (cell == 'B' || cell == 'W')
                {
                    borders.Add(cell);
                }
            }
        }
        Owner owner = Owner.None;
        if (borders.Count == 1)
        {
            owner = borders.Contains('B') ? Owner.Black : Owner.White;
        }
        return Tuple.Create(owner, visited);
    }

    public Dictionary<Owner, HashSet<(int, int)>> Territories()
    {
        var result = new Dictionary<Owner, HashSet<(int, int)>>()
        {
            [Owner.Black] = [],
            [Owner.White] = [],
            [Owner.None] = []
        };
        var seen = new HashSet<(int, int)>();
        for (int x = 0; x < width; x++)
        {
            for (int y = 0; y < height; y++)
            {
                if (board[x, y] == ' ' && !seen.Contains((x, y)))
                {
                    var (owner, coords) = Territory((x, y));
                    foreach (var coord in coords)
                        seen.Add(coord);
                    result[owner].UnionWith(coords);
                }
            }
        }
        return result;
    }

    private bool InBounds(int x, int y)
    => x >= 0 && y >= 0 && x < width && y < height;
    private static IEnumerable<(int,int)> Neighbors(int x, int y)
    {
        yield return (x - 1, y);
        yield return (x + 1, y);
        yield return (x, y - 1);
        yield return (x, y + 1);
    }
}

File: javascript/go-counting/go-counting.js
//
// This is only a SKELETON file for the 'Go Counting' exercise. It's been provided as a
// convenience to get you started writing code faster.
//
const EMPTY = ' ';
const BLACK = 'BLACK';
const WHITE = 'WHITE';
const NONE = 'NONE';

export class GoCounting {
  constructor(board) {
    this.board = board.map(row => [...row]);
    this.height = this.board.length;
    this.width = this.board[0]?.length || 0;
  }

  getTerritory(x, y) {
    if (x < 0 || x >= this.width || y < 0 || y >= this.height) {
      return { error: 'Invalid coordinate'};
    }
    if (this.board[y][x] !== EMPTY) {
      return { owner: NONE, territory: [] };
    }

    const visited = new Set();
    const queue = [[x, y]];
    visited.add(`${x},${y}`);

    const territory = [];
    const stonesTouched = new Set();

    while (queue.length > 0) {
      const [cx, cy] = queue.shift();
      territory.push([cx, cy]);

      const neighbors = [[cx + 1, cy], [cx - 1, cy], [cx, cy + 1], [cx, cy - 1]];

      for (const [nx, ny] of neighbors) {
        if (nx < 0 || nx >= this.width || ny < 0 || ny >= this.height) continue;

        const cell = this.board[ny][nx];
        if (cell === EMPTY) {
          const key = `${nx},${ny}`;
          if (!visited.has(key)) {
            visited.add(key);
            queue.push([nx, ny]);
          }
        } else {
          stonesTouched.add(cell);
        }
      }
    }

    let owner = NONE;
    if (stonesTouched.size === 1) {
      owner = stonesTouched.has('B') ? BLACK : WHITE;
    }

    territory.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
    return { owner, territory };
  }

  getTerritories() {
    const summary = { territoryBlack: [], territoryWhite: [], territoryNone: [] };
    const globalVisited = new Set();

    for (let y = 0; y < this.height; y++) {
      for (let x = 0; x < this.width; x++) {
        if (this.board[y][x] === EMPTY && !globalVisited.has(`${x},${y}`)) {
          const { owner, territory } = this.getTerritory(x, y);

          for (const [tx, ty] of territory) {
            globalVisited.add(`${tx},${ty}`);
          }

          if (owner === BLACK) summary.territoryBlack.push(...territory);
          else if (owner === WHITE) summary.territoryWhite.push(...territory);
          else summary.territoryNone.push(...territory);
        }
      }
    }
    summary.territoryBlack.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
    summary.territoryWhite.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
    summary.territoryNone.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
    return summary;
  }
}
